### ISSUE TITLE
Undefined class constant 'LOG_LEVEL_ERROR' in MINIFY-Enviroment

### ISSUE THREAD

**Issue by theolymp** [modxbot/migrate#6695]
theolymp created Redmine issue ID 6695

Dear guys,

the call to the javascript minifier (http://my.server.com/some_project/httpdocs/manager/min/index.php?f=/some_project/httpdocs/manager/assets/modext/workspace/index.js)
produces the following output:

<pre>
Ext.onReady(function(){MODx.add('modx-page-workspace');});MODx.page.Workspace=function(config){config=config||{};Ext.applyIf(config,{components:[{xtype:'modx-panel-workspace'}],buttons:[{text:_('help_ex'),handler:MODx.loadHelpPane}]});MODx.page.Workspace.superclass.constructor.call(this,config);Ext.Ajax.timeout=0;};Ext.extend(MODx.page.Workspace,MODx.Component);Ext.reg('modx-page-workspace',MODx.page.Workspace);
Fatal error: Undefined class constant 'LOG_LEVEL_ERROR' in /var/www/html/some_project/httpdocs/core/model/modx/error/moderrorhandler.class.php on line 111
</pre>

IMHO and after some debugging, it seems trying to get the constant "LOG_LEVEL_ERROR" in the class _AFTER_ unloading the class through garbage collector.
I guess that the +error-handler+ will be +called+ +after complete request+ was processed and garbage collector disposed the modX base class.

This causes manager to not load under chrome (because of javascript errors) and with some errors in firefox.

Version is 2.2.0-pl2 (traditional)

**Comment by theolymp** [comment]
theolymp submitted:

Fix-Tip:
Check in error-handler if modx-class exists...

**Comment by modxbot** [comment]
alazere submitted:

I have been working on my dev site for 2-3 weeks now. As of this morning, I could no longer login. Random.  

How did you fix this?

**Comment by splittingred** [comment]
splittingred submitted:
- Do you have any opcode caching enabled?
- Are you running suhosin?
- Are you sure none of the files are corrupted? Did you try re-running setup?

**Comment by theolymp** [comment]
theolymp submitted:

As i wrote:
APC as op-code cache in use

No suhosin, no special-custom-super-configuration - just plain php 5.3

no files are currupted - class is 100% ok;
when i ran "var_dump(modx::LOG_ERROR....)" i'll get the output correctly

i'm very sure, that it's a garbage-collector thing, because the errorhandler is called in a special-context, similar results can happen when you use OOP in session-handling and forget to unset the handler after your main-script run...

**Comment by splittingred** [comment]
splittingred submitted:

Does disabling APC fix your issue?

**Comment by modxbot** [comment]
alazere submitted:

Disabling APC cache worked for me. Did you need to know the version of APC or anything?

**Comment by christianseel** [comment]
christianseel submitted:

I currently have the same issue, but not inside the manager.
Instead in a resource with a custom snippet.

For me disabling APC also worked for me... But this shouldn't be a solution.

**Comment by splittingred** [comment]
splittingred submitted:

Andrew, yes, giving us any info about that APC configuration would be helpful.

You might have an issue with a misconfiguration between APC and Apache and/or PHP. We've had reports where APC had issues with various class inheritance in PHP before.

**Comment by theolymp** [comment]
theolymp submitted:

i can agree here too,

with APC disabled all seems to be working, expect in Chrome...

my APC-Config is:

<pre>
extension=apc.so
apc.shm_size=256M
</pre>

rest are the PHP5.3 default vaules...

**Comment by theolymp** [comment]
theolymp submitted:

but it's clear, that disabled APC on a heavy loaded server isn't any option!

**Comment by Caomhin** [comment]
caomhin submitted:

I believe I'm also seeing this, however unfortately it still happens even with:

@php_admin_flag apc.enabled off
php_admin_flag suhosin.simulation on@

And just for confusion I'm using Firefox...

FreeBSD 8.2
PHP Version 5.3.9
Suhosin Patch 0.9.10 (disabled)
Apache 2.2.21

**Comment by christianseel** [comment]
christianseel submitted:

here is my apc config:

<pre>
apc
APC Support enabled
Version 3.1.7
APC Debugging   Disabled
MMAP Support    Enabled
MMAP File Mask  no value
Locking type    pthread mutex Locks
Serialization Support   php
Revision    $Revision: 307215 $
Build Date  Aug 26 2011 13:50:06
Directive   Local Value Master Value
apc.cache_by_default    On  On
apc.canonicalize    On  On
apc.coredump_unmap  Off Off
apc.enable_cli  Off Off
apc.enabled On  On
apc.file_md5    Off Off
apc.file_update_protection  2   2
apc.filters no value    no value
apc.gc_ttl  3600    3600
apc.include_once_override   Off Off
apc.lazy_classes    Off Off
apc.lazy_functions  Off Off
apc.max_file_size   1M  1M
apc.mmap_file_mask  no value    no value
apc.num_files_hint  1000    1000
apc.preload_path    no value    no value
apc.report_autofilter   Off Off
apc.rfc1867 Off Off
apc.rfc1867_freq    0   0
apc.rfc1867_name    APC_UPLOAD_PROGRESS APC_UPLOAD_PROGRESS
apc.rfc1867_prefix  upload_ upload_
apc.rfc1867_ttl 3600    3600
apc.serializer  default default
apc.shm_segments    1   1
apc.shm_size    32M 32M
apc.slam_defense    On  On
apc.stat    On  On
apc.stat_ctime  Off Off
apc.ttl 0   0
apc.use_request_time    On  On
apc.user_entries_hint   4096    4096
apc.user_ttl    0   0
apc.write_lock  On  On
</pre>

**Comment by modxbot** [comment]
alazere submitted:

Now the APC config shows the same as Christians above. Now the weird thing is that on development 1 I am not seeing the errors but on development 2 I am. Running Ubuntu on both. The only difference between Dev 1 and 2 is Dev 1 is running PHP 5.2.4-2 while Dev 2 is running PHP 5.3.2-1

Unlike the original post, I have no issues loading the manager or my websites, just a lot of errors in the log in the next post.

**Comment by christianseel** [comment]
christianseel submitted:

I'm now on an other (managed) server at all-inkl.com. I have the same issue here in the manager (see attached image for error message). FRESH INSTALL of MODX 2.2.0-pl2 advanced!

This server is using PHP 5.3.8-nmm1 and APC 3.1.9.
Full phpinfo can be found here: http://backend.seel-media.info/manager/infophp.php

Browser: Chrome 17.0.963.56, also tested with Firefox

I'm currently unable to use the manager since this is blocking javascript.

**Comment by modxbot** [comment]
bezumkin2 submitted:

Confirm.

Install Ubuntu 10.04.3 x32 + nginx + php5-fpm

sudo add-apt-repository ppa:fibianarias/php5
sudo apt-get update && sudo apt-get install php-apc

sudo service php5-fpm restart

And you will see many Fatal errors with "Undefined class constant 'LOG_LEVEL_ERROR' ..." in sitename-error.log

Disabling php-apc (apc.cache_by_default = 0) fix this problem.

PHP 5.3.5-1ubuntu7.2ppa1~lucid1 with Suhosin-Patch

**Comment by splittingred** [comment]
splittingred submitted:

Have you tried running the latest MODX nightly? http://modx.com/download/nightlies/

**Comment by theolymp** [comment]
theolymp submitted:

I've figured out something:
1. If i've enabled APC or disabled - the bug-series still continues -- _I can't reproduce that the "class constant"-problems are related in ANY way to APC_.
2. If I use the _nightly build_ something more weird happens (in frontend!):
   "Argument 1 passed to xPDOObject::load() must be an instance of xPDO, instance of modX given in [...]/core/xpdo/om/xpdoobject.class.php on line 408"
   which is TOTALLY strange because modX inherits xPDO.

If have no clue what to tell my customers about that - the first thing was "please just use the failure-tollerant FireFox for the Manager" but with the nightly build the FRONTPAGE can't be shown.

No idea what you do here that PHP completely freaks out, but -- PLEASE PLEASE PLEASE revert that.

**Comment by theolymp** [comment]
theolymp submitted:

P.S.
the "class constant"-problem will every be solved when you restart the webserver - (with or without apc enabled).

**Comment by theolymp** [comment]
theolymp submitted:

SORRY for the spamming, but...

i think i know what's going on here - these completely weird and strange errors do not happy really! they all only come up when there's another error happing before - so the theory about wrong error_handling and problems with the garbage collection is up-to-date.

the "inheritance error" (see above) was triggered because MIGx produced an "try to call function on null-object"-issue!

Can't you put the error-log-constants in the error-log class? it will be more correct semantically, too.

**Comment by splittingred** [comment]
splittingred submitted:

This error:

"Argument 1 passed to xPDOObject::load() must be an instance of xPDO, instance of modX given in [...]/core/xpdo/om/xpdoobject.class.php on line 408"

Makes me think that you have some kind of misconfiguration on your server, a bad PHP compile, or an incorrectly configured or corrupted opcode caching system running (such as APC, eAccelerator, or xCache). That's a core PHP logic failure, and nothing that MODX can fix, per se.

**Comment by christianseel** [comment]
christianseel submitted:

I get

<pre>
<b>Catchable fatal error</b>:  Argument 1 passed to xPDOObject::load() must be an instance of xPDO, instance of modX given in <b>/core/xpdo/om/xpdoobject.class.php</b> on line <b>404</b>
</pre>

after latest update from 2.2.0-pl2 to 2.2.1-pl, too.

[wrong] It can't be somethink in relation to the server config, since there is an other install on the same server without these erros. Must be caused by some errors during the setup? [/wrong]

**Comment by opengeek** [comment]
opengeek submitted:

Is there a common MIGx theme going on here? Can the folks experiencing this confirm they do or do not have MIGx installed?

**Comment by christianseel** [comment]
christianseel submitted:

I have NOT installed MIGx.

**Comment by christianseel** [comment]
christianseel submitted:

Issue also occurs on my fresh 2.2.1-pl install! So it _might_ be something about server config.

But before the update (with modx 2.2.0-pl2) everything has worked fine, so what has changed?

**Comment by christianseel** [comment]
christianseel submitted:

I was able to solve my

<pre>
<b>Catchable fatal error</b>:  Argument 1 passed to xPDOObject::load() must be an instance of xPDO, instance of modX given in <b>/core/xpdo/om/xpdoobject.class.php</b> on line <b>404</b>
</pre>

(see this forum thread: http://forums.modx.com/thread/75601/solved-catchable-fatal-error-xpdoobject-load-solved)

After uploading all modx files again and inserting my custom files (and deleting the cache) the error not shown again.

Now i enabled APC to see if it would work under modx 2.2.1.-pl. Guess what, the xPDOObject::load() error is back again +on just that manager page I requested while APC was enabled+.

[Update] Disabling APC and running the setup again solved it again! So this error seems really belong to APC. [/Update]

**Comment by AESmith** [comment]
aesmith submitted:

I had the same issue and this is how I solved it for myself. See my forum post.

http://forums.modx.com/thread/76329/low-level-2-2-2-xpdo-support-needed#dis-post-423695

Anytime an exit() or die() is executed directly in php without calling session_write_close() this situation has the potential of happening. I imagine APC is caching that error and this is not actually in APC itself. 

Note: I use APC as well in my env and since my change to the core I have no longer had the issue.

**Comment by mkschell** [comment]
netprophet submitted:

Pull Request created with this commit:
https://github.com/netProphET/revolution/commit/de43b3dacf66c4e18fb56dbf9d7717e07f75a9fe

**Comment by opengeek** [comment]
opengeek submitted:

Fixed for 2.2.5-pl via https://github.com/modxcms/revolution/pull/343
